src/vcf/validator.cpp

```diff
@@ -150,38 +164,76 @@ namespace ebi
             return false;
         }
         std::unique_ptr<Parser> validator = build_parser(sourceName, validationLevel, version);
-        return validate(line, input, *validator, outputs);
+        return validate(line, uncompressed_input, *validator, outputs);
     }
 
-    bool is_compressed_file(const std::string &source,
-                            const std::vector<char> &line)
+    void create_uncompressed_stream(std::istream & input,
+                                    const std::string & sourceName,
+                                    boost::iostreams::filtering_istream & uncompressed_input)
     {
-        if (source != ebi::vcf::STDIN && is_compressed_extension(source)) {
-            return true;
+        std::vector<char> line;
+        get_magic_num(input, line);
+
+        std::string file_ext = get_compression(sourceName, line);
+        check_readability_of_file(file_ext);
+
+        if(file_ext == BZ2) {
+            uncompressed_input.push(boost::iostreams::bzip2_decompressor());
+        } else if(file_ext == TAR_GZ) {
+            uncompressed_input.push(boost::iostreams::gzip_decompressor());
         }
-        return is_compressed_magic_num(line);
+
+        uncompressed_input.push(input);
     }
 
-    bool is_compressed_extension(std::string const & source)
+    void get_magic_num(std::istream & stream, std::vector<char> & container)
+    {
+        char c;
+        int i=0;
+        container.clear();
+
+        while (stream && stream.get(c)) {
+            container.push_back(c);
+            i++;
+            if (c == '\n' || i > 4) break;
+        }
+
+        for (int j = 0; j < (int)container.size(); ++j) {
+            stream.unget();
```

Comment:
@jmmut I have added a check to throw if stream fails.
Reason why I am not preferring reading first line is that in case of a compressed stream the size of `first_line`  can be very large. I think function `read_line` will almost read the whole stream in one go, if the stream is compressed.
The `first_line` may not be used directly as it may be compressed(in case of compressed stream). We will again have to send it for filtering. That will be equivalent to the approach I proposed earlier in beginning of this PR.

Reply:
not exactly.

- In case it is not compressed, you will read either the first 4 bytes of the "##fileformat" line or until "\n" if the line is invalid and shorter than 4 bytes, and that string should be used as-is. In fact, if you find a "\n" in those 4 bytes the `detect_version` function should throw an exception later.
- In case it is compressed, you will read the 4 bytes and only need the subsequent bytes to send them through the filtering_stream. In this case, we should not read until "\n" is found because if it's compressed, the newlines won't appear as such.